File: 2025/python/day_8/part_1.py
import sys
from .common import (
    FILE_PATH,
    Point,
    Segment,
    parse_line_into_point,
    parse_points_into_segments,
    sort_segments_by_length,
    island_sizes,
)


def connect_pairs(segments: list[Segment], num_of_connections: int) -> None:
    for i in range(num_of_connections):
        segment = segments[i]
        segment.start.connect(segment.end)


def sort_islands_by_size(island_sizes: list[int]) -> list[int]:
    return sorted(island_sizes, reverse=True)


def product_largest_islands(island_sizes: list[int], num: int) -> int:
    product = 1
    for i in range(num):
        size = island_sizes[i]
        product *= size
    return product


def day8_part1() -> None:
    try:
        running_id = 0
        points = []
        
        with open(FILE_PATH, "r") as f:
            for line in f:
                line = line.strip()
                point = parse_line_into_point(line, running_id)
                points.append(point)
                running_id += 1
        
        segments = sort_segments_by_length(parse_points_into_segments(points))
        connect_pairs(segments, 1000)
        islands = sort_islands_by_size(island_sizes(points))
        solution = product_largest_islands(islands, 3)
        
        print(f"Solution: {solution}")
    
    except FileNotFoundError:
        print("File not found")
        sys.exit(1)
    except ValueError as e:
        print(f"Parsing error: {e}")
        sys.exit(1)
    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)

File: 2025/ts/src/day-8/part-1.ts
import * as fs from 'fs';
import * as readline from 'readline';
import {
  PATH,
  Point,
  Segment,
  parseLineIntoPoint,
  parsePointsIntoSegments,
  sortSegmentsByLength,
  islandSizes,
} from './common';

function connectPairs(segments: Segment[], numOfConnections: number): void {
  for (let i = 0; i < numOfConnections; i++) {
    const segment = segments[i];
    segment.start.connect(segment.end);
  }
}

function sortIslandsBySize(islandSizes: number[]): number[] {
  return islandSizes.sort((a, b) => b - a);
}

function productLargestIslands(islandSizes: number[], num: number): number {
  let product = 1;
  for (let i = 0; i < num; i++) {
    const size = islandSizes[i];
    product *= size;
  }
  return product;
}

export async function day8Part1(): Promise<void> {
  try {
    const fileStream = fs.createReadStream(PATH);
    const rl = readline.createInterface({
      input: fileStream,
      crlfDelay: Infinity,
    });

    let runningId = 0;
    const points: Point[] = [];

    for await (const line of rl) {
      const trimmedLine = line.trim();
      const point = parseLineIntoPoint(trimmedLine, runningId);
      points.push(point);
      runningId++;
    }

    const segments = sortSegmentsByLength(parsePointsIntoSegments(points));
    connectPairs(segments, 1000);
    const islands = sortIslandsBySize(islandSizes(points));
    const solution = productLargestIslands(islands, 3);

    console.log(`Solution: ${solution}`);
  } catch (error) {
    console.error('Error:', error);
    process.exit(1);
  }
}
